nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/StandardPythonBridge.java

```diff
@@ -0,0 +1,315 @@
+            controllerProcess = new PythonProcess(processConfig, serviceTypeLookup, envHome);
+            controllerProcess.start();
+            running = true;
+        } catch (final Exception e) {
+            shutdown();
+            throw e;
+        }
+    }
+
+
+    @Override
+    public void discoverExtensions() {
+        ensureStarted();
+        final List<String> extensionsDirs = processConfig.getPythonExtensionsDirectories().stream()
+            .map(File::getAbsolutePath)
+            .collect(Collectors.toList());
+        final String workDirPath = processConfig.getPythonWorkingDirectory().getAbsolutePath();
+        controllerProcess.getController().discoverExtensions(extensionsDirs, workDirPath);
+    }
+
+    @Override
+    public PythonProcessorBridge createProcessor(final String identifier, final String type, final String version, final boolean preferIsolatedProcess) {
+        ensureStarted();
+
+        logger.debug("Creating Python Processor of type {}", type);
+
+        final PythonProcess pythonProcess = getProcessForNextComponent(type, version, preferIsolatedProcess);
+        final String workDirPath = processConfig.getPythonWorkingDirectory().getAbsolutePath();
+
+        final PythonController controller = pythonProcess.getController();
+        final PythonProcessorAdapter processorAdapter = controller.createProcessor(type, version, workDirPath);
+        final PythonProcessorBridge processorBridge = new StandardPythonProcessorBridge.Builder()
+            .controller(controller)
+            .processorAdapter(processorAdapter)
+            .processorType(type)
+            .processorVersion(version)
+            .workingDirectory(processConfig.getPythonWorkingDirectory())
+            .moduleFile(new File(controller.getModuleFile(type, version)))
+            .build();
+
+        pythonProcess.addProcessor(identifier, preferIsolatedProcess);
+        final ExtensionId extensionId = new ExtensionId(type, version);
+        processorCountByType.merge(extensionId, 1, Integer::sum);
+        return processorBridge;
+    }
+
+    @Override
+    public synchronized void onProcessorRemoved(final String identifier, final String type, final String version) {
+        final ExtensionId extensionId = new ExtensionId(type, version);
+        final List<PythonProcess> processes = processesByProcessorType.get(extensionId);
+        if (processes == null) {
+            return;
+        }
+
+        // Find the Python Process that has the Processor, if any, and remove it.
+        // If there are no additional Processors in the Python Process, remove it from our list and shut down the process.
+        final Iterator<PythonProcess> processItr = processes.iterator(); // Use iter so we can call remove()
+        while (processItr.hasNext()) {
+            final PythonProcess process = processItr.next();
+            final boolean removed = process.removeProcessor(identifier);
```

Comment:
I added a python processor without cofigure it and remove it from the canvas. But the started python process is not stopped. I debued the code, and it seems, removeProcessor is called with a wrong value?

<img width="943" alt="image" src="https://user-images.githubusercontent.com/227213/224340612-418d7b7d-d365-45fd-a169-706daed5ee01.png">

As you can see on the picture identifier is different what is inside processorPrefersIsolation map. Of course this is not a serious bug, because it will be stop during nifi shutdown

Reply:
I think this is working correctly, actually.
In order to determine some details about Processors, NiFi needs an instance of the Processor. Things like what properties are available, default values, etc. It uses this information, among other things, for building the documentation, etc.
So that's why you see the temp-component-PrettyPrintJson there. That's the temporary component. If you then add a PrettyPrintJson processor, it will add it to that same Process. Because it requires virtually no resources, we combine the temp component along with the actual Processor in order to avoid creating more processes than we have to. But we can't eliminate that Python process while the temp processor still exists.
So basically once you add a Processor of type X, there will remain a Python Process for it until NiFi is restarted, even if you remove the Processor.
Perhaps this could be improved in the future, but it's a pretty low priority at the moment, given that we don't expect an exorbitant number of different Processor types and definitely don't expect an exorbitant number types of Processors added to the canvas and removed.